src/Setup_geometry.jl

```diff
@@ -1177,3 +1179,354 @@ function Compute_ThermalStructure(Temp, X, Y, Z, Phase, s::LinearWeightedTempera
+It returns Bn -> which is the point B coordinate in the new transformed system.
+"""
+function transform_coordinate!(X,Y,Z,XT,YT,A,B,direction)
+
+    # find the slope of AB points
+    s = (B[2]-A[2])/(B[1]-A[1])
+
+    angle_rot = -atand(s);
+
+    # Shift the origin
+    XT .= X .-A[1];
+    YT .= Y .-A[2];
+
+    Bn = zeros(3);
+
+    Bn[1] = B[1]-A[1];
+
+    Bn[2] = B[2]-A[2];
+
+    Bn[3] = 0.0;
+
+    # Transform the coordinates
+    Rot3D!(XT,YT,Z, angle_rot, 0);
+
+    YT .= YT*direction;
+
+    #Find Point B in the new coordinate system
+    roty = [cosd(-0)          0              sind(-0) ; 0 1 0 ; -sind(-0) 0  cosd(-0)];
+    rotz = [cosd(angle_rot) -sind(angle_rot)        0 ; sind(angle_rot) cosd(angle_rot) 0 ; 0 0 1]
+
+    Bn = rotz*Bn;
+    Bn = roty*Bn;
+
+    return Bn
+end
+
+"""
+    find_slab!(X,Y,Z,d,ls,θ_max,A,B,Top,Bottom,seg_slab,D0,L0)
+
+Function that finds the slab. It loops over the Top and Bottom surface of the slab.
+It creates small polygons where and check if in the transformed coordinate system there are particles belonging to that polygon
+then interpolates the distance from the top, and the current length from the corners.
+"""
+function find_slab!(X,Y,Z,d,ls,θ_max,A,B,Top,Bottom,seg_slab,D0,L0,direction)
+
+    # Create the XT,YT
+    XT = zero(X);
+    YT = zero(Y);
+
+    # Function to transform the coordinate
+    xb = transform_coordinate!(X,Y,Z,XT,YT,A,B,direction);
+
+    # dl
+    dl = L0/seg_slab;
+
+    l = 0  # length at the trench position
+
+    # Construct the slab
+    for i = 1:(seg_slab-1)
+
+        ln = l+dl;
+
+        pa = (Top[i,1],Top[i,2]); # D = 0 | L = l
+
+        pb = (Bottom[i,1],Bottom[i,2]); # D = -D0 | L=l
+
+        pc = (Bottom[i+1,1],Bottom[i+1,2]); # D = -D0 |L=L+dl
+
+        pd = (Top[i+1,1],Top[i+1,2]) # D = 0| L = L+dl
+
+        # Create the polygon
+        poly_y = [pa[1],pb[1],pc[1],pd[1]];
+
+        poly_z = [pa[2],pb[2],pc[2],pd[2]];
+
+        # find a sub set of particles
+        ymin = minimum(poly_y);
+
+        ymax = maximum(poly_y);
+
+        zmin = minimum(poly_z);
+
+        zmax = maximum(poly_z);
+
+        ind_s = findall(0.0.<= XT.<= xb[1] .&& ymin .<= YT .<= ymax .&& zmin .<= Z .<= zmax);
+
+        # Find the particles
+        yp = YT[ind_s];
+
+        zp = Z[ind_s];
+
+        # Initialize the ind that are going to be used by inpoly
+        ind = zeros(Bool,size(zp));
+
+        # inPoly! [Written by Arne Spang, must be updated with the new version]
+        inPolygon!(ind,poly_y,poly_z,yp,zp);
+
+        # indexes of the segment
+        ind_seg = ind_s[ind]
+
+        # Prepare the variable to interpolate {I put here because to allow also a variation of thickness of the slab}
+        D = [0.0,-D0,-D0,0.0];
+
+        L = [l,l,ln,ln];
+
+        # Interpolations
+        points = [pa[1] pa[2];pb[1] pb[2];pc[1] pc[2];pd[1] pd[2]]'
```

Comment:
Here I have a question. Because I did not grasp fully the concept of staticarrays. Are the static arrays immutable? Or do I can change the values? Because in this case I can convert these three arrays in static array and change the value during the slab geometry identification.

Reply:
There are both mutable `MArray` and static `SArray` arrays, where you can modify the value of the elements of the first one. But `SArray` are allocation free and typically faster, so sometimes may be better to "allocate" a new small `SArray', instead of  using a `MArray`. The only condition to use any of them is that you need to know their length at compile time, which you already know.